### DISCUSSION TITLE
How can I update the same state field as subscribed to without triggering an infinite loop?

### DISCUSSION THREAD

**Discussion by ViktorAtterlonn** [pmndrs/zustand#2036] [Q&A] [answered]
Hello,

I hope you're all doing well. I'm currently facing an issue while using useStore.subscribe in my project, and I'm seeking some guidance to resolve it.

I've set up a useStore.subscribe function to monitor changes to a specific field in my store, and within the subscribe callback, I'm performing some asynchronous operations and updating the same field using the set function. However, this seems to trigger an infinite loop, causing unexpected behavior in my application.

Here's a simplified example of my code:

```typescript
const { set, get } = useStore.getState()

useStore.subscribe(
  (state) => state.listOfItems,
  async (listOfItems, prev) => {
    const worker = getWorkerInstance()

    const updatedListOfItems = await worker.populateFieldsWithAsyncData(listOfItems)

    set((state) => {
      state.listOfItems = updatedListOfItems
    })
  }
)
```

I've tried various approaches to mitigate this issue, but I haven't been successful so far. I suspect I might be missing something crucial regarding how the subscribe function should be used in such scenarios.

I would greatly appreciate it if someone could provide some insights or guidance on how to avoid this infinite loop while achieving my intended functionality. Any help or suggestions would be highly appreciated.

Thank you in advance for your assistance!
Viktor

**Comment by dai-shi** [comment, marked as answer]
It depends on your requirement.
You want to solve the following simplified scenario first.
```js
function onListOfItemsChanged(currentListOfItems) {
  const callback = async (listOfItems, prev) => {
    const worker = getWorkerInstance()
    const updatedListOfItems = await worker.populateFieldsWithAsyncData(listOfItems)
    onListOfItemsChanged(updatedListOfItems)
  }
  callback(currentListOfItems)
)
```

**Reply by ViktorAtterlonn** [reply]
Got it! To clarify, I should implement a diff check in the callback and abort if the values are the same. That should certainly solve my problem. I thought there might be some functionality in Zustand that supports intercepting state changes and making alterations before setting them.

**Reply by dai-shi** [reply]
Sounds good.
No, there's nothing special in Zustand.
You should be able to read the code as it's pretty short: https://twitter.com/dai_shi/status/1583082766081531905

**Reply by ViktorAtterlonn** [reply]
I'll continue on that route. Thanks for your help and for building this awesome state manager! 😊
